File: examples/read_plates.py
#!/usr/bin/env python3
"""read_plates.py - a two-stage license-plate reader built on pycamtrt.

Stage 1 ("detect") runs a YOLOv8 plate detector over each stream; stage 2
("read") crops every plate detection and runs an LPRNet-style CTC OCR
engine over the crop. This is the showcase cascade example for the
pycamtrt v0 API - see python/pycamtrt/__init__.py for the full API this
is built from.

Note on output text: the bundled LPRNet engine (models/lprnet_b1-32_fp32_sm86.engine)
was trained on a Chinese license-plate charset, so decoded strings on
non-Chinese plates will look garbled (e.g. "<su>A123..."). That's expected -
this example exercises the detect -> crop -> OCR *mechanics*, not charset
fit for any particular region.

Run it (from the repo root, inside the tensorrt-dev docker container, with
the Python-dev conda env providing the interpreter):

    docker run --rm --gpus all --network host \\
        -e NVIDIA_DRIVER_CAPABILITIES=compute,utility,video \\
        -v $(pwd):/workspace \\
        -v ~/anaconda3:/home/<user>/anaconda3 \\
        -w /workspace tensorrt-dev bash -c '
            ln -sf libnvcuvid.so.1 /usr/lib/x86_64-linux-gnu/libnvcuvid.so
            PYTHONPATH=/workspace/build:/workspace/python \\
            ~/anaconda3/envs/Python-dev/bin/python3 \\
            examples/read_plates.py rtsp://localhost:8554/cam1 rtsp://localhost:8554/cam2
        '

The RTSP sources come from tools/stream_farm/farm.sh. To serve plate
footage instead of the default webcam loop, override CLIP before bringing
the farm up:

    cd tools/stream_farm && CLIP=media/atlas_plate_g30.mp4 ./farm.sh up 2

Pass --verify to also run the C++ side's per-frame CPU-reference check on
the detector (pycamtrt.Pipeline(verify=True)); every consumed result's
verify_ok is then asserted True and a "PYTHON VERIFY N/N" line is printed
at the end instead of the plate summary.
"""
import sys
from collections import Counter

import pycamtrt

DETECT_ENGINE = "models/yolov8n_plates_b1-16_fp16_sm86.engine"
READ_ENGINE = "models/lprnet_b1-32_fp32_sm86.engine"

# skip=2 (infer every other decoded frame) + max_frames=600 (decoded frames
# per stream) -> ~300 inferred frames/stream, ~20s wall at 30fps source.
SKIP = 2
MAX_FRAMES = 600


def build_pipeline(urls, verify):
    streams = pycamtrt.Streams(urls)

    detect = pycamtrt.Layer("detect")
    eng = detect.add(pycamtrt.Engine(streams, DETECT_ENGINE))
    detect.add(pycamtrt.Postprocess(eng, family="yolo", score=0.4))

    read = pycamtrt.Layer("read")
    # Cross-layer edge: an Engine fed a Postprocess step auto-crops each
    # detection from the earlier layer before running the OCR engine on it.
    oeng = read.add(pycamtrt.Engine(detect.steps[-1], READ_ENGINE))
    read.add(pycamtrt.Postprocess(oeng, family="ctc"))

    return pycamtrt.Pipeline(
        streams, layers=[detect, read], skip=SKIP, max_frames=MAX_FRAMES,
        verify=verify,
    )


def main():
    args = sys.argv[1:]
    verify = "--verify" in args
    urls = [a for a in args if a != "--verify"] or ["rtsp://localhost:8554/cam1"]

    frames_seen = 0
    plates_seen = 0
    plate_counts = Counter()
    verify_ok = 0
    verify_total = 0

    with build_pipeline(urls, verify) as pipe:
        for r in pipe:
            frames_seen += 1
            texts = r.outputs["read"]
            if texts:
                plates_seen += len(texts)
                plate_counts.update(texts)
                for t in texts:
                    print(f"s{r.stream_id} frame {r.frame_no}: {t}")
            if verify:
                verify_total += 1
                if r.verify_ok:
                    verify_ok += 1

        n_streams = len(urls)
        per_stream = []
        for i in range(n_streams):
            si = pipe.get_stream_info(i)
            per_stream.append(si)

    print("\n---- summary ----")
    print(f"frames seen: {frames_seen}  plates read: {plates_seen}")
    print("distinct plate strings:")
    for text, count in plate_counts.most_common():
        print(f"  {count:5d}  {text!r}")
    for i, si in enumerate(per_stream):
        print(f"s{i}: decoded={si.decoded} reconnects={si.reconnects}")

    if verify:
        print(f"PYTHON VERIFY {verify_ok}/{verify_total}")
        if verify_ok != verify_total:
            sys.exit(1)


if __name__ == "__main__":
    main()

File: examples/read_plates/read_plates.cpp
// read_plates.cpp - native C++ port of read_plates.py, built directly on
// pycamtrt::Pipeline (src/core/graph.h/pipeline.h/result.h) instead of the
// pycamtrt Python compiler layer. Same two-stage step graph (detect ->
// crop -> OCR), same per-result plate prints, same summary block - see
// examples/read_plates/README.md for run commands and expected output.
//
// Note on output text (same caveat as the Python side): the bundled
// LPRNet engine was trained on a Chinese license-plate charset, so
// decoded strings on non-Chinese plates look garbled - expected, this
// exercises the detect -> crop -> OCR mechanics, not charset fit.
//
// Usage:
//   read_plates URL [URL...] [--verify]

#include <algorithm>
#include <cstdio>
#include <cstring>
#include <map>
#include <memory>
#include <string>
#include <vector>

#include "core/pipeline.h"

using pycamtrt::Family;
using pycamtrt::FrameResult;
using pycamtrt::LayerDesc;
using pycamtrt::Pipeline;
using pycamtrt::PipelineConfig;
using pycamtrt::StepDesc;
using pycamtrt::StepKind;
using pycamtrt::StreamDesc;
using pycamtrt::StreamInfo;

namespace {

const char* const kDetectEngine = "models/yolov8n_plates_b1-16_fp16_sm86.engine";
const char* const kReadEngine = "models/lprnet_b1-32_fp32_sm86.engine";

// skip=2 (infer every other decoded frame) + max_frames=600 (decoded
// frames per stream) -> ~300 inferred frames/stream, ~20s wall at 30fps
// source - same cadence as the Python example.
const int kSkip = 2;
const int kMaxFrames = 600;

// Builds the same detect(yolo) -> read(ctc) step graph
// pycamtrt.Layer/Engine/Postprocess compiles to in read_plates.py's
// build_pipeline().
PipelineConfig BuildConfig(const std::vector<std::string>& urls, bool verify) {
    PipelineConfig cfg;
    cfg.streams.reserve(urls.size());
    for (const std::string& u : urls) cfg.streams.push_back(StreamDesc{u});
    cfg.skip = kSkip;
    cfg.max_frames = kMaxFrames;
    cfg.verify = verify;
    cfg.log = [](const std::string& s) { printf("%s\n", s.c_str()); };

    StepDesc engine_step;
    engine_step.kind = StepKind::Engine;
    engine_step.input = -1;  // raw stream source
    engine_step.engine_path = kDetectEngine;
    const int engine_idx = (int)cfg.steps.size();
    cfg.steps.push_back(engine_step);

    StepDesc post_step;
    post_step.kind = StepKind::Postprocess;
    post_step.input = engine_idx;
    post_step.family = Family::YoloDetect;
    post_step.score_thresh = 0.4f;
    const int post_idx = (int)cfg.steps.size();
    cfg.steps.push_back(post_step);

    LayerDesc detect;
    detect.name = "detect";
    detect.steps = {engine_idx, post_idx};
    cfg.layers.push_back(detect);

    // Cross-layer edge: an Engine fed a Postprocess step from an earlier
    // layer auto-crops each detection before running the OCR engine on it
    // (same mechanism the Python Engine(detect.steps[-1], READ_ENGINE) call
    // compiles down to - see pipeline.cpp's cascade-crop path).
    StepDesc read_engine_step;
    read_engine_step.kind = StepKind::Engine;
    read_engine_step.input = post_idx;
    read_engine_step.engine_path = kReadEngine;
    const int read_engine_idx = (int)cfg.steps.size();
    cfg.steps.push_back(read_engine_step);

    StepDesc read_post_step;
    read_post_step.kind = StepKind::Postprocess;
    read_post_step.input = read_engine_idx;
    read_post_step.family = Family::Ctc;
    const int read_post_idx = (int)cfg.steps.size();
    cfg.steps.push_back(read_post_step);

    LayerDesc read;
    read.name = "read";
    read.steps = {read_engine_idx, read_post_idx};
    cfg.layers.push_back(read);

    return cfg;
}

}  // namespace

int main(int argc, char** argv) {
    std::vector<std::string> urls;
    bool verify = false;
    for (int i = 1; i < argc; i++) {
        if (!std::strcmp(argv[i], "--verify")) verify = true;
        else urls.push_back(argv[i]);
    }
    if (urls.empty()) urls.push_back("rtsp://localhost:8554/cam1");

    std::unique_ptr<Pipeline> pipeline;
    try {
        pipeline.reset(new Pipeline(BuildConfig(urls, verify)));
    } catch (const std::exception& e) {
        fprintf(stderr, "pipeline init failed: %s\n", e.what());
        return -1;
    }
    pipeline->Start();

    int frames_seen = 0, plates_seen = 0;
    int verify_ok = 0, verify_total = 0;
    std::map<std::string, int> plate_counts;

    for (;;) {
        FrameResult r;
        const Pipeline::PollStatus st = pipeline->Poll(&r, 500);
        if (st == Pipeline::PollStatus::Timeout) continue;
        if (st == Pipeline::PollStatus::Finished) break;

        frames_seen++;
        // r.texts is the "read" layer's back-compat field (first Ctc
        // child), aligned with r.detections - exactly what the Python
        // side's r.outputs["read"] resolves to for this single-child
        // cascade (see core/result.h's WHY-comment).
        if (!r.texts.empty()) {
            plates_seen += (int)r.texts.size();
            for (const std::string& t : r.texts) {
                plate_counts[t]++;
                printf("s%d frame %d: %s\n", r.stream_id, r.frame_no, t.c_str());
            }
        }
        if (verify) {
            verify_total++;
            if (r.verify_ok) verify_ok++;
        }
    }

    const int n_streams = (int)urls.size();
    std::vector<StreamInfo> per_stream;
    per_stream.reserve(n_streams);
    for (int i = 0; i < n_streams; i++) per_stream.push_back(pipeline->GetStreamInfo(i));
    pipeline.reset();  // Stop() + teardown before printing (same order as
                       // the Python example's `with` block exit)

    printf("\n---- summary ----\n");
    printf("frames seen: %d  plates read: %d\n", frames_seen, plates_seen);
    printf("distinct plate strings:\n");
    std::vector<std::pair<std::string, int>> sorted_counts(plate_counts.begin(),
                                                            plate_counts.end());
    std::sort(sorted_counts.begin(), sorted_counts.end(),
              [](const auto& a, const auto& b) { return a.second > b.second; });
    for (const auto& [text, count] : sorted_counts) {
        printf("  %5d  '%s'\n", count, text.c_str());
    }
    for (int i = 0; i < n_streams; i++) {
        printf("s%d: decoded=%d reconnects=%d\n", i, per_stream[i].decoded,
               per_stream[i].reconnects);
    }

    if (verify) {
        printf("VERIFY %d/%d\n", verify_ok, verify_total);
        if (verify_ok != verify_total) return 1;
    }
    return 0;
}
